File: src/modules/algorithm/seed/algorithms/topological-sort/languages/python/bfs-solution.py
from typing import List
from collections import deque

def topological_sort(n: int, edges: List[List[int]]) -> List[int]:
    """
    Topological Sort using Kahn's Algorithm (BFS)

    Args:
        n: The number of vertices in the graph.
        edges: The edges of the graph represented as a list of [u, v] pairs.

    Returns:
        A topological ordering of the vertices.
    """
    graph = {i: [] for i in range(n)}
    in_degree = [0] * n

    for u, v in edges:
        graph[u].append(v)
        in_degree[v] += 1

    queue = deque([i for i in range(n) if in_degree[i] == 0])
    result = []
    count = 0

    while queue:
        u = queue.popleft()
        result.append(u)
        count += 1

        for v in graph[u]:
            in_degree[v] -= 1
            if in_degree[v] == 0:
                queue.append(v)

    return result if count == n else []  # Check for cycle 

File: src/modules/algorithm/seed/algorithms/topological-sort/languages/javascript/bfs-solution.js
/**
 * Topological Sort using Kahn's Algorithm (BFS)
 *
 * @param {number} n - The number of vertices in the graph.
 * @param {number[][]} edges - The edges of the graph represented as an array of [u, v] pairs.
 * @returns {number[]} A topological ordering of the vertices.
 */
function topologicalSort(n, edges) {
  const graph = {};
  const inDegree = new Array(n).fill(0);

  for (let i = 0; i < n; i++) {
    graph[i] = [];
  }

  for (const [u, v] of edges) {
    graph[u].push(v);
    inDegree[v]++;
  }

  const queue = [];
  for (let i = 0; i < n; i++) {
    if (inDegree[i] === 0) {
      queue.push(i);
    }
  }

  const result = [];
  let count = 0;

  while (queue.length > 0) {
    const u = queue.shift();
    result.push(u);
    count++;

    for (const v of graph[u]) {
      inDegree[v]--;
      if (inDegree[v] === 0) {
        queue.push(v);
      }
    }
  }

  return count === n ? result : []; // Check for cycle
}

module.exports = { topologicalSort };
